File: frontend/src/api/claims.ts
/**
 * Copyright 2026 SciLoop contributors
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * 证据链与 Claim 数据层（WP13-T6）：封装 WP13 的三个端点。
 *
 * 分层约定（contracts.code_style.frontend）：组件内**禁止**直接 fetch，
 * 一切请求经 `src/api/*.ts`；写操作由 `client.ts` 自动带 `X-Owner-Token`。
 *
 * 覆盖端点（附录 B.5）：
 *   GET  /evidence/{type}/{id}        解析多态证据为可跳转结构（附录 E.3）
 *   GET  /evidence/types              受控值域 + fulltext_gate 阈值
 *   GET  /drafts/{id}/claims          Claim 列表（三态 + 证据数 + 覆盖率统计）
 *   POST /drafts/{id}/verify-evidence 重跑校验，返回 unsupported_spans 与覆盖率（Owner）
 */
import { get, post, type RequestOptions } from './client'

/** contracts.enums.claim_status */
export type ClaimStatus = 'supported' | 'contradicted' | 'insufficient'

/** contracts.evidence_rules.evidence_types */
export type EvidenceType =
  | 'paper_span'
  | 'card_field'
  | 'experiment_run'
  | 'experiment_passport'
  | 'decision'

/** ``verify_span`` 三 verdict（哈希优先） */
export type EvidenceVerdict = 'valid' | 'valid_by_hash' | 'invalid'

/** ``paper_cards`` 的 8 字段（card_field 证据的取值白名单） */
export const CARD_FIELDS = [
  'research_problem',
  'core_method',
  'key_innovation',
  'technical_route',
  'experimental_setup',
  'main_conclusions',
  'limitations',
  'transferable',
] as const

/** evidence_id 语义：evidences.id 或该类型自身主键 */
export type EvidenceIdKind = 'evidence' | 'native'

/** ``evidences`` 行 + 附录 E.3 解析结果（含平铺别名，见 api/v1/evidence.py:_flatten） */
export interface EvidenceDetail {
  /** 平铺别名，等同 evidence_id */
  id?: number | null
  evidence_id: number
  evidence_type: EvidenceType | string
  owner_type?: string | null
  owner_id?: number | null
  weight?: number | null
  /** 论文摘要；字段缺失一律 null，不编造 */
  paper?: {
    id?: number | null
    title?: string | null
    venue?: string | null
    venue_source?: string | null
    published_at?: string | null
    doi?: string | null
    source?: string | null
    external_id?: string | null
  } | null
  paper_id?: number | null
  /** 定位必须携带 document_version（禁止只用裸字符偏移） */
  document_version?: string | null
  span?: {
    paper_span_id?: number
    section_name?: string | null
    page_number?: number | null
    bbox?: Record<string, number> | null
    char_start?: number | null
    char_end?: number | null
  } | null
  card?: {
    card_id?: number | null
    version?: number | null
    field?: string | null
    value?: string | null
    value_available?: boolean
    unavailable_reason?: string | null
  } | null
  quote_text?: string | null
  quote_sha256?: string | null
  /** 平铺别名：verification.verdict */
  verdict?: EvidenceVerdict | string | null
  verification?: {
    verdict?: EvidenceVerdict | string
    hash_match?: boolean
    offset_match?: boolean | null
    expected_quote_sha256?: string
    stored_quote_sha256?: string
    reason?: string
  } | null
  /** fulltext_gate 如实披露 */
  gate?: {
    ok?: boolean
    parse_status?: string | null
    coverage?: number | null
    reason?: string | null
    coverage_note?: string | null
    evidence_scope?: string | null
  } | null
  /** 平铺别名：gate.coverage */
  coverage?: number | null
  /** 平铺别名：gate.ok */
  spans_allowed?: boolean
  section_name?: string | null
  page_number?: number | null
  char_start?: number | null
  char_end?: number | null
  evidence_scope?: string | null
  jump_url?: string | null
  warnings?: string[]
  id_kind?: EvidenceIdKind
  [key: string]: unknown
}

/** ``GET /drafts/{id}/claims`` 的单条 Claim */
export interface DraftClaim {
  id: number
  claim_id: number
  draft_id: number
  section_heading?: string | null
  claim_text?: string | null
  /** ``draft_claims`` 无文本列时的等价别名 */
  text?: string | null
  is_factual: boolean
  support_status: ClaimStatus | string
  /** ``support_status`` 的前端别名（ClaimBadge / DraftViewer 用） */
  status: ClaimStatus | string
  status_reason?: string | null
  evidence_count: number
  evidence_ids: number[]
  evidence: Array<Record<string, unknown>>
  created_at?: string | null
  [key: string]: unknown
}

export interface DraftClaimCounts {
  supported: number
  contradicted: number
  insufficient: number
  factual: number
  non_factual?: number
  total: number
}

export interface DraftClaimsResponse {
  draft_id: number
  /** false = paper_drafts 尚无该行（WP14 未产出）：如实未就绪，不是 404 */
  draft_found: boolean
  items: DraftClaim[]
  total: number
  page: number
  page_size: number
  counts: DraftClaimCounts
  /** supported ÷ 事实性总数；事实性总数为 0 时为 null（不用 0/1 冒充） */
  claim_coverage: number | null
  claim_coverage_persisted?: number | null
  coverage_formula?: string
  coverage_note?: string | null
  unsupported_spans: UnsupportedSpan[]
  warnings: string[]
  note?: string | null
}

/** 事实性但非 supported 的段落（供前端整段高亮） */
export interface UnsupportedSpan {
  claim_id: number | null
  index?: number | null
  section_heading?: string | null
  claim_text?: string | null
  char_start?: number | null
  char_end?: number | null
  support_status: ClaimStatus | string
  status_reason?: string | null
  evidence_count?: number
}

export interface VerifyEvidenceRequest {
  /** 调试入口：草稿未落库时按此正文直接校验（只读，不写库） */
  content_md?: string | null
  persist?: boolean
  use_llm_conflict?: boolean
  max_claims?: number
}

export interface VerifyEvidenceResponse {
  draft_id: number | null
  draft_found: boolean
  persisted: boolean
  /** ``draft_persisted``（读库校验）或 ``content_md_debug``（纯文本只读校验） */
  mode?: string
  claim_coverage: number | null
  counts: Record<string, number>
  claims: Array<Record<string, unknown>>
  unsupported_spans: UnsupportedSpan[]
  unsupported: UnsupportedSpan[]
  unsupported_total?: number
  citation_map: Record<string, string>
  citation_count?: number
  citation_source?: string
  conflict_detection?: string
  coverage_formula?: string
  coverage_note?: string | null
  warnings: string[]
  verified_at?: string
  [key: string]: unknown
}

/** ``ClaimBadge`` 点击后向父组件回传的载荷 */
export interface ClaimSelectPayload {
  claim_id: number | null
  status: string
  is_factual: boolean
  evidence_count: number
  evidence_ids: number[]
  char_start: number | null
  char_end: number | null
  reason: string | null
}

/** 读一条证据（附录 E.3）。`idKind='native'` 时 id 为该类型自身主键。 */
export function getEvidenceDetail(
  evidenceType: EvidenceType | string,
  evidenceId: number | string,
  options: { idKind?: EvidenceIdKind; cardField?: string; signal?: AbortSignal } = {},
) {
  const query: RequestOptions['query'] = {}
  if (options.idKind && options.idKind !== 'evidence') query.id_kind = options.idKind
  if (options.cardField && (evidenceType === 'card_field' && options.idKind === 'native')) {
    query.card_field = options.cardField
  }
  return get<EvidenceDetail>(`/evidence/${encodeURIComponent(evidenceType)}/${evidenceId}`, {
    query,
    signal: options.signal,
  })
}

/** 受控证据类型 + fulltext_gate 阈值（前端不硬编码 0.60） */
export function getEvidenceTypes(signal?: AbortSignal) {
  return get<{
    evidence_types: EvidenceType[]
    id_kinds: EvidenceIdKind[]
    fulltext_gate: { parse_status: string; min_coverage: number; rule: string }
  }>('/evidence/types', { signal })
}

/** Claim 列表（草稿不存在时 draft_found=false，不是错误） */
export function getDraftClaims(
  draftId: number | string,
  options: {
    supportStatus?: ClaimStatus
    factualOnly?: boolean
    page?: number
    pageSize?: number
    signal?: AbortSignal
  } = {},
) {
  const query: RequestOptions['query'] = {}
  if (options.supportStatus) query.support_status = options.supportStatus
  if (options.factualOnly) query.factual_only = true
  if (options.page) query.page = options.page
  if (options.pageSize) query.page_size = options.pageSize
  return get<DraftClaimsResponse>(`/drafts/${draftId}/claims`, { query, signal: options.signal })
}

/** 重跑 Claim 级证据校验（Owner 面；返回 unsupported_spans 与 claim_coverage） */
export function verifyDraftEvidence(
  draftId: number | string,
  body: VerifyEvidenceRequest = {},
  signal?: AbortSignal,
) {
  return post<VerifyEvidenceResponse>(`/drafts/${draftId}/verify-evidence`, {
    body,
    signal,
    timeoutMs: 120_000,
  })
}

/** 本地按 Claim 列表复算覆盖率（与后端同一公式，用于交叉核对，不做权威来源） */
export function computeCoverageFromClaims(items: DraftClaim[]): number | null {
  const factual = items.filter((item) => item.is_factual)
  if (factual.length === 0) return null
  const supported = factual.filter((item) => item.status === 'supported').length
  return Math.round((supported / factual.length) * 1000) / 1000
}

/** 覆盖率展示：null → 「无定义」，绝不显示 0 */
export function formatCoverage(value: number | null | undefined): string {
  if (value === null || value === undefined || Number.isNaN(value)) return '无定义'
  return `${Math.round(value * 1000) / 10}%`
}

File: server/api/v1/claims.py
# Copyright 2026 SciLoop contributors
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""草稿 Claim 级证据端点（WP13-T5，附录 B.5）。

==========================================  ==========================================
``GET  /drafts/{id}/claims``                Claim 列表（三态 + 证据数 + 覆盖率统计）
``POST /drafts/{id}/verify-evidence``       重跑校验，返回 ``unsupported_spans`` 与覆盖率
==========================================  ==========================================

口径与硬约束
------------
- ``claim_coverage = supported Claim 数 ÷ 事实性 Claim 总数``；事实性总数为 0 时返回
  ``null`` 并附 ``coverage_note``（**不用 0 或 1 冒充**）。
- 事实性 Claim 必给三态 ``supported | contradicted | insufficient``，不得留空；
  非 supported 的事实性 Claim 一律出现在 ``unsupported_spans`` 里（供前端高亮）。
- **容忍草稿不存在**：``paper_drafts`` 由 WP14 并行产出，行不存在时两个端点都返回
  ``draft_found=false`` + ``warnings``，**不抛 404**，UI 可显示如实的未就绪态。
- ``POST`` 支持 ``content_md`` 调试入口：草稿未落库时可直接校验纯文本（只读，不写库）。
- 写操作仅 Owner 面（contracts.api_contract：「所有 POST 默认要求 X-Owner-Token」；
  ``public_demo`` 只放行 GET），非 Owner 一律 403。
"""

from __future__ import annotations

import logging
from collections.abc import AsyncIterator, Mapping, Sequence
from typing import Annotated, Any

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel, Field
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from core.security import require_owner
from db.session import AsyncSessionLocal
from services.evidence import (
    CLAIM_STATUSES,
    EvidenceError,
    claim_coverage,
    load_draft,
    make_llm_conflict_detector,
    verify_claims,
    verify_content,
)

logger = logging.getLogger("sciloop.wp13.claims_api")

router = APIRouter(tags=["claims"])

MAX_PAGE_SIZE = 500
DEFAULT_PAGE_SIZE = 200
STATUS_HELP = "受控 Claim 状态之一：" + " / ".join(CLAIM_STATUSES)

DRAFT_MISSING_NOTE = (
    "paper_drafts 中不存在该草稿（WP14 尚未产出）：返回未就绪态，未伪造任何 Claim"
)


async def _session() -> AsyncIterator[AsyncSession]:
    if AsyncSessionLocal is None:  # pragma: no cover - 部署期驱动缺失
        raise _error(
            status.HTTP_503_SERVICE_UNAVAILABLE,
            "database_unavailable",
            "异步数据库会话工厂不可用（DATABASE_URL / asyncpg 未就绪）",
        )
    async with AsyncSessionLocal() as session:
        yield session


DbSession = Annotated[AsyncSession, Depends(_session)]


def _error(status_code: int, code: str, message: str, detail: Any = None) -> HTTPException:
    return HTTPException(
        status_code=status_code,
        detail={"code": code, "message": message, "detail": detail},
    )


def _num(value: Any) -> float | None:
    if value is None:
        return None
    try:
        return float(value)
    except (TypeError, ValueError):  # pragma: no cover - NUMERIC 列受控
        return None


# --------------------------------------------------------------------------------------
# GET /drafts/{id}/claims
# --------------------------------------------------------------------------------------
async def _claims_with_evidence(
    session: AsyncSession, draft_id: int
) -> tuple[list[Mapping[str, Any]], dict[int, list[dict[str, Any]]]]:
    """读 ``draft_claims`` 及其绑定的 ``evidences``（一次查询，避免 N+1）。"""
    rows = (
        await session.execute(
            text(
                """
                SELECT id, draft_id, section_heading, claim_text, is_factual,
                       support_status, status_reason, evidence_count, created_at
                  FROM draft_claims
                 WHERE draft_id = :draft_id
                 ORDER BY id
                """
            ),
            {"draft_id": int(draft_id)},
        )
    ).mappings().all()
    claims = [dict(row) for row in rows]
    if not claims:
        return claims, {}

    evidence_rows = (
        await session.execute(
            text(
                """
                SELECT id, owner_id, evidence_type, paper_span_id, paper_id,
                       card_field, experiment_run_id, experiment_passport_id,
                       decision_log_id, metric_name, quote_text, weight
                  FROM evidences
                 WHERE owner_type = 'draft_claim' AND owner_id = ANY(:claim_ids)
                 ORDER BY id
                """
            ),
            {"claim_ids": [int(row["id"]) for row in claims]},
        )
    ).mappings().all()

    grouped: dict[int, list[dict[str, Any]]] = {}
    for row in evidence_rows:
        grouped.setdefault(int(row["owner_id"]), []).append(
            {
                "id": int(row["id"]),
                "evidence_id": int(row["id"]),
                "evidence_type": row["evidence_type"],
                "paper_id": row["paper_id"],
                "paper_span_id": row["paper_span_id"],
                "card_field": row["card_field"],
                "experiment_run_id": row["experiment_run_id"],
                "experiment_passport_id": row["experiment_passport_id"],
                "decision_log_id": row["decision_log_id"],
                "metric_name": row["metric_name"],
                "quote_text": row["quote_text"],
                "weight": _num(row["weight"]),
            }
        )
    return claims, grouped


def _claim_item(
    row: Mapping[str, Any], evidence: Sequence[Mapping[str, Any]]
) -> dict[str, Any]:
    """单条 Claim 的响应体（``status`` 为 ``support_status`` 的前端别名）。"""
    support_status = str(row.get("support_status") or "insufficient")
    return {
        "id": int(row["id"]),
        "claim_id": int(row["id"]),
        "draft_id": int(row["draft_id"]),
        "index": None,
        "section_heading": row.get("section_heading"),
        "claim_text": row.get("claim_text"),
        "text": row.get("claim_text"),
        "is_factual": bool(row.get("is_factual")),
        "support_status": support_status,
        "status": support_status,
        "status_reason": row.get("status_reason"),
        "evidence_count": int(row.get("evidence_count") or 0),
        "evidence_ids": [int(item["id"]) for item in evidence],
        "evidence": [dict(item) for item in evidence],
        "created_at": row["created_at"].isoformat() if row.get("created_at") else None,
    }


@router.get(
    "/drafts/{draft_id}/claims",
    summary="Claim 列表（supported / contradicted / insufficient + 证据数与覆盖率）",
)
async def list_draft_claims(
    draft_id: int,
    session: DbSession,
    support_status: str | None = Query(None, description=STATUS_HELP),
    factual_only: bool = Query(
        False, description="只返回事实性 Claim（claim_coverage 的分母口径）"
    ),
    page: int = Query(1, ge=1),
    page_size: int = Query(DEFAULT_PAGE_SIZE, ge=1, le=MAX_PAGE_SIZE),
) -> dict[str, Any]:
    """返回该草稿的 Claim 列表与统计；草稿不存在时返回未就绪态（非 404）。"""
    if support_status is not None and support_status not in CLAIM_STATUSES:
        raise _error(
            status.HTTP_400_BAD_REQUEST,
            "invalid_claim_status",
            f"support_status='{support_status}' 不在受控值域内",
            {"allowed": list(CLAIM_STATUSES)},
        )

    draft = await load_draft(session, int(draft_id))
    if draft is None:
        logger.info("list_draft_claims 草稿不存在 draft_id=%s", draft_id)
        return {
            "draft_id": int(draft_id),
            "draft_found": False,
            "items": [],
            "total": 0,
            "page": page,
            "page_size": page_size,
            "counts": dict.fromkeys(CLAIM_STATUSES, 0) | {"total": 0, "factual": 0},
            "claim_coverage": None,
            "coverage_note": "草稿不存在：claim_coverage 无定义",
            "unsupported_spans": [],
            "warnings": [DRAFT_MISSING_NOTE],
            "note": DRAFT_MISSING_NOTE,
        }

    claims, grouped = await _claims_with_evidence(session, int(draft_id))
    items = [_claim_item(row, grouped.get(int(row["id"]), [])) for row in claims]
    if factual_only:
        items = [item for item in items if item["is_factual"]]
    if support_status is not None:
        items = [item for item in items if item["support_status"] == support_status]

    factual_all = [
        item for item in items if item["is_factual"]
    ] if not factual_only else items
    counts = dict.fromkeys(CLAIM_STATUSES, 0)
    for item in factual_all:
        counts[item["support_status"]] = counts.get(item["support_status"], 0) + 1
    totals = {
        **counts,
        "factual": len(factual_all),
        "non_factual": 0 if factual_only else sum(1 for item in items if not item["is_factual"]),
        "total": len(items),
    }
    supported = counts["supported"]
    coverage = claim_coverage(supported, len(factual_all))

    total = len(items)
    start = (page - 1) * page_size
    page_items = items[start : start + page_size]

    return {
        "draft_id": int(draft_id),
        "draft_found": True,
        "items": page_items,
        "total": total,
        "page": page,
        "page_size": page_size,
        "counts": totals,
        "claim_coverage": coverage,
        "claim_coverage_persisted": _num(draft.get("claim_coverage")),
        "coverage_formula": "supported Claim 数 ÷ 事实性 Claim 总数",
        "coverage_note": (
            None
            if coverage is not None
            else "无事实性 Claim：claim_coverage 无定义（不用 0 或 1 冒充）"
        ),
        "unsupported_spans": [
            {
                "claim_id": item["claim_id"],
                "section_heading": item["section_heading"],
                "claim_text": item["claim_text"],
                "support_status": item["support_status"],
                "status_reason": item["status_reason"],
            }
            for item in factual_all
            if item["support_status"] != "supported"
        ],
        "warnings": [],
        "note": (
            "使用 GET /drafts/{id}/claims 读取落库状态；POST /drafts/{id}/verify-evidence "
            "会重跑拆分与哈希优先校验并回写三态与 claim_coverage"
        ),
    }


# --------------------------------------------------------------------------------------
# POST /drafts/{id}/verify-evidence
# --------------------------------------------------------------------------------------
class VerifyEvidenceRequest(BaseModel):
    """重跑校验的入参（全部可选）。"""

    content_md: str | None = Field(
        default=None,
        description=(
            "调试入口：草稿尚未落库（WP14 未产出）时按此正文直接校验（只读，不写库）。"
            "不传则读 paper_drafts.content_md"
        ),
    )
    persist: bool = Field(
        default=True,
        description="是否把三态写回 draft_claims 并回写 paper_drafts.claim_coverage",
    )
    use_llm_conflict: bool = Field(
        default=False,
        description=(
            "是否启用 LLM 冲突判定（WP02 chat_json）。默认关闭：规则层三态可离线复现；"
            "启用后必须给出冲突证据 id"
        ),
    )
    max_claims: int = Field(default=0, ge=0, description="0 = 不限；>0 按出现顺序截断")


@router.post(
    "/drafts/{draft_id}/verify-evidence",
    summary="重跑 Claim 级证据校验（返回 unsupported_spans 与 claim_coverage，owner 面）",
    dependencies=[Depends(require_owner)],
)
async def verify_draft_evidence(
    draft_id: int,
    session: DbSession,
    body: VerifyEvidenceRequest | None = None,
) -> dict[str, Any]:
    """重跑拆分 + 哈希优先校验 + 三态判定，返回未支撑段落位置与覆盖率。

    草稿行不存在时：若给了 ``content_md`` 则走只读调试校验；否则返回
    ``draft_found=false`` 与 ``warnings``（**不抛 404**，容忍 WP14 并行未就绪）。
    """
    request = body or VerifyEvidenceRequest()
    conflict_detector = (
        make_llm_conflict_detector() if request.use_llm_conflict else None
    )
    try:
        if request.content_md is not None:
            report = await verify_content(
                request.content_md,
                session=session,
                draft_id=int(draft_id),
                conflict_detector=conflict_detector,
                max_claims=request.max_claims,
            )
            report["mode"] = "content_md_debug"
        else:
            report = await verify_claims(
                int(draft_id),
                session=session,
                conflict_detector=conflict_detector,
                persist=request.persist,
                max_claims=request.max_claims,
            )
            report["mode"] = "draft_persisted"
    except EvidenceError as exc:
        raise _error(
            status.HTTP_422_UNPROCESSABLE_ENTITY, exc.code, exc.message, exc.detail
        ) from exc

    report["coverage_formula"] = "supported Claim 数 ÷ 事实性 Claim 总数"
    report["unsupported_total"] = len(report.get("unsupported_spans") or [])
    logger.info(
        "verify_evidence draft_id=%s mode=%s draft_found=%s coverage=%s unsupported=%s",
        draft_id,
        report.get("mode"),
        report.get("draft_found"),
        report.get("claim_coverage"),
        report.get("unsupported_total"),
    )
    return report
